EOS-25688: Reduce mkfs time for multiple io services

suvratjoshi opened the pull request:
Problem:
On a given node for all the motr processes (confds and ioservices),
presently, mkfs is executed serially via Hare mini-provisioner. It is
executed in parallel across the nodes but not on a node. So ioservices
mkfs does not complete before confd's mkfs completes. Now we don't
require confds to be running for ioservices mkfs (this is done by
passing confd.xc to motr-mkfs). Thus, all can be executed in parallel
on a given node.
For sequential execution of mkfs, it takes significant amount of time
if the number of mkfs requests are more (typically when we have
multinode clusters). With this parallel execution of mkfs we need
to reduce this time.

Solution:
The idea here is to make the mkfs commands to run in parallel.
This PR addresses this using concurrent process execution
mechanism.

Signed-off-by: Suvrat Joshi <suvrat.joshi@seagate.com>

mssawant commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -389,10 +401,27 @@ def start_mkfs(hostname: str, hare_config_dir: str):
+    finish = perf_counter()
+    perf_result = str(finish - start)
+
+    if (-1 in ret_values):
```
-1 seems too specific, why not check `if 0 not in ret__values:`?

suvratjoshi replied:
Hi @mssawant ,
"if 0 not in" is not suitable here as the list can have 0 and -1s.
I have not done this change.
Of course it is possible to do this either by 
1. having an iterative check method in the list (but will not be optimal)
2. OR we can have True/False return values from start_mkfs and simply use it in the if condition (but I did not prefer it as we generally have not done like that I guess)

mssawant replied:
Right, instead of `if 0 not in` we need to use `if not all(rc == 0 for rc in result)`

suvratjoshi replied:
ok. Done!

andriytk commented:
Even now, m0mkfs run in parallel: 1st for all confds, and then for all ios Motr services. It's because m0mkfs for ios requires confds running already.

Do you mean it's required to run them all in parallel? How?

suvratjoshi commented:
> Do you mean it's required to run them all in parallel? How?

Hi @andriytk 
This is for parallel invocation of mkfs commands.

andriytk commented:
> Hi @andriytk This is for parallel invocation of mkfs commands.

But should be invoked in parallel already on all the nodes, isn't it so now?

suvratjoshi commented:
Testing Done :
A build for this PR (with # ) was created http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1164/ and its docker image deployed on 4 node cluster successfully :
```
[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2444 /]# hctl status
Data pool:
    # fid name
    0x6f00000000000001:0xab 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x10b 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2447
    [started]  hax        0x7200000000000001:0x31  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2447@2001
    [started]  ioservice  0x7200000000000001:0x34  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2447@3001
    [started]  ioservice  0x7200000000000001:0x41  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2447@3002
    [started]  confd      0x7200000000000001:0x4e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2447@3003
    [started]  s3server   0x7200000000000001:0x51  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2447@4001
    [started]  s3server   0x7200000000000001:0x54  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2447@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2444  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2444@2001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2444@3001
    [started]  ioservice  0x7200000000000001:0x17  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2444@3002
    [started]  confd      0x7200000000000001:0x24  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2444@3003
    [started]  s3server   0x7200000000000001:0x27  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2444@4001
    [started]  s3server   0x7200000000000001:0x2a  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2444@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2446
    [started]  hax        0x7200000000000001:0x5b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2446@2001
    [started]  ioservice  0x7200000000000001:0x5e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2446@3001
    [started]  ioservice  0x7200000000000001:0x6b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2446@3002
    [started]  confd      0x7200000000000001:0x78  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2446@3003
    [started]  s3server   0x7200000000000001:0x7b  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2446@4001
    [started]  s3server   0x7200000000000001:0x7e  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2446@4002
    cortx-data-headless-svc-ssc-vm-g3-rhev4-2445
    [started]  hax        0x7200000000000001:0x85  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2445@2001
    [started]  ioservice  0x7200000000000001:0x88  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2445@3001
    [started]  ioservice  0x7200000000000001:0x95  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2445@3002
    [started]  confd      0x7200000000000001:0xa2  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2445@3003
    [started]  s3server   0x7200000000000001:0xa5  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2445@4001
    [started]  s3server   0x7200000000000001:0xa8  inet:tcp:cortx-data-headless-svc-ssc-vm-g3-rhev4-2445@4002
[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2444 /]#
```

suvratjoshi commented:
For measuring mkfs execution time I have added some logs in this PR, which will be outputted to file, inside the pod : 
`/etc/cortx/log/hare/log/{machine_id}/hare_deployment/setup.log'

These timings are checked after successful deployment :
```

[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2444 /]# vi /etc/cortx/log/hare/log/00784f357c24486f889930d102fa7f79/hare_deployment/setup.log
2021-11-19 14:12:12,876 [INFO] fid = 0x7200000000000001:0x24

2021-11-19 14:12:12,882 [INFO] fid = 0x7200000000000001:0xa

2021-11-19 14:12:12,883 [INFO] fid = 0x7200000000000001:0x17

2021-11-19 14:12:43,577 [INFO] Total time taken for all mkfs on this node = 30.710139208997134

[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2445 /]# vi /etc/cortx/log/hare/log/b8f4acd233d84e0585e15d6b1102e549/hare_deployment/setup.log
2021-11-19 14:12:29,729 [INFO] fid = 0x7200000000000001:0xa2

2021-11-19 14:12:29,737 [INFO] fid = 0x7200000000000001:0x88

2021-11-19 14:12:29,737 [INFO] fid = 0x7200000000000001:0x95

2021-11-19 14:12:51,758 [INFO] Total time taken for all mkfs on this node = 22.084245572972577

[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2446 /]# vi /etc/cortx/log/hare/log/8bfb0446400849ebb178d61a1cd5d78b/hare_deployment/setup.log
2021-11-19 14:12:41,325 [INFO] fid = 0x7200000000000001:0x78

2021-11-19 14:12:41,390 [INFO] fid = 0x7200000000000001:0x5e

2021-11-19 14:12:41,390 [INFO] fid = 0x7200000000000001:0x6b

2021-11-19 14:13:18,290 [INFO] Total time taken for all mkfs on this node = 36.98659632098861

[root@cortx-data-headless-svc-ssc-vm-g3-rhev4-2447 /]# vi /etc/cortx/log/hare/log/6f6dcc4ff3d246e78395b0e754c705c5/hare_deployment/setup.log
2021-11-19 14:12:33,870 [INFO] fid = 0x7200000000000001:0x4e

2021-11-19 14:12:33,878 [INFO] fid = 0x7200000000000001:0x34

2021-11-19 14:12:33,879 [INFO] fid = 0x7200000000000001:0x41

2021-11-19 14:13:14,605 [INFO] Total time taken for all mkfs on this node = 40.744305767002515

```
Average is approx 32 seconds.

Similarly, I had also checked the time required for mkfs execution, on a code without having this PR's changes.
(http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1145/)

The Average mkfs execution time without code changes was approx = 55 seconds

suvratjoshi commented:
After addressing review comment, I have rebased the patch, and have retriggered a build for deployment :
http://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.9/job/cortx-all-image-custom-ci/1166/

mssawant commented:
@andriytk, on a given node for all the motr processes (confds and ioservices), presently, mkfs is executed serially via Hare mini-provisioner. It is executed in parallel across the nodes but not on a node. So ioservices mkfs does not complete before confd's mkfs completes. Now we don't require confds to be running for ioservices mkfs (this is done by passing confd.xc to motr-mkfs). Thus, all can be executed in parallel on a given node.

mssawant commented:
looks good, tested 1164 15 node cluster, provisioning time has improved.. (when captured the data pods were getting provisioned already).
```
cortx-data-provisioner-pod-ssc-vm-g2-rhev4-1630        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g2-rhev4-1631        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g2-rhev4-1632        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g2-rhev4-1635        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g2-rhev4-2237        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g2-rhev4-2238        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2107        0/1     Completed           0          14m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2184        0/1     Completed           0          14m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2198        0/1     Completed           0          14m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2278        0/1     Completed           0          14m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2279        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2281        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2282        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2283        0/1     Completed           0          13m
cortx-data-provisioner-pod-ssc-vm-g3-rhev4-2284        0/1     Completed           0          13m
```

andriytk commented:
> @andriytk, on a given node for all the motr processes (confds and ioservices), presently, mkfs is executed serially via Hare mini-provisioner. It is executed in parallel across the nodes but not on a node. So ioservices mkfs does not complete before confd's mkfs completes. Now we don't require confds to be running for ioservices mkfs (this is done by passing confd.xc to motr-mkfs). Thus, all can be executed in parallel on a given node.

Got it. Thanks, @mssawant. It would be useful to mention this in the description above, I think.

suvratjoshi commented:
> Got it. Thanks, @mssawant. It would be useful to mention this in the description above, I think.

@andriytk,
Updated. Thanks you!

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -389,10 +401,27 @@ def start_mkfs(hostname: str, hare_config_dir: str):
         for fid in svc_fids:
-            execute([cmd, fid, '--conf'])
+            # real command line
+            logging.info(f'fid = {str(fid)}\n')
```
```suggestion
            logging.info('Running mkfs [fid = %s]', fid)
```

knekrasov replied:
BTW this logging should happen within `start_mkfs()` function. Because at this point the work is planned only. There is no guarantee that this work will be processed immediately.

knekrasov replied:
Addressed.

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -389,10 +401,27 @@ def start_mkfs(hostname: str, hare_config_dir: str):
     cmd = '/usr/libexec/cortx-motr/motr-mkfs'
     # start mkfs for confd, ios services
+    start = perf_counter()
+    cmd_list = []
```
Type hint is needed.

knekrasov replied:
Addressed

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -378,7 +380,17 @@ def start(args):
+        execute(cmd_line)
+        rc = 0
+    except Exception as error:
+        logging.error('Error during mkfs start (%s)', error)
```
It is more beneficial if the error describes which mkfs process actually failed (at least, which fid was it).

knekrasov replied:
Addressed.

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -389,10 +401,27 @@ def start_mkfs(hostname: str, hare_config_dir: str):
+    with concurrent.futures.ProcessPoolExecutor() as executor:
+        ret_values = list(executor.map(start_mkfs, cmd_list))
+    finish = perf_counter()
+    perf_result = str(finish - start)
```
```suggestion
    perf_result = finish - start
```

knekrasov replied:
Applied.

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -389,10 +401,27 @@ def start_mkfs(hostname: str, hare_config_dir: str):
+    perf_result = str(finish - start)
+
+    if not all(rc == 0 for rc in ret_values):
+        logging.error('Error during mkfs start')
```
```suggestion
        logging.error('One or more mkfs processes failed to start. Please check the logs above for details.')
```

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -378,7 +380,17 @@ def start(args):
+    except Exception as error:
+        logging.error('Error during mkfs start (%s)', error)
+        rc = -1
+    return(rc)
```
```suggestion
    return rc
```

knekrasov replied:
Done

knekrasov commented on provisioning/miniprov/hare_mp/main.py:
```diff
@@ -389,10 +401,27 @@ def start_mkfs(hostname: str, hare_config_dir: str):
+    perf_result = str(finish - start)
+
+    if not all(rc == 0 for rc in ret_values):
+        logging.error('Error during mkfs start')
```
Don't we need to raise an exception here?

The aim is to make sure that the exit code is not 0 if there was a problem during the execution.

knekrasov replied:
Done

mssawant commented:
Merging as deployment is working fine with the patch, post deployment data pods restarts were seen due to setup related issues.
